Update documentation for Reflection Abstract Factory

froschdesign opened the pull request:
The documentation for Reflection Abstract Factory has been updated to improve the clarity and flow of the information. The changes include rewording for better understanding, reordering of sections for better logical progression, and additional warnings and recommendations concerning performance issues and alternatives.

gsteel commented on docs/book/v4/reflection-abstract-factory.md:
```diff
@@ -1,67 +1,80 @@
-either an abstract factory, or mapped to specific service names as a factory:
+Writing a factory class for each and every service that has dependencies can be tedious, particularly in early development as you are still sorting out dependencies.
+
+To alleviate this issue during development, laminas-servicemanager ships with the `ReflectionBasedAbstractFactory`, which provides a [reflection-based approach](https://www.php.net/manual/en/intro.reflection.php) to instantiation, resolving constructor dependencies to the relevant services.
```
```suggestion
To alleviate this issue during development, laminas-servicemanager ships with the `ReflectionBasedAbstractFactory`, which provides a [reflection-based approach](https://www.php.net/manual/intro.reflection.php) to instantiation, resolving constructor dependencies to the relevant services.
```

froschdesign replied:
Good catch! 👍🏻

gsteel commented on docs/book/v4/reflection-abstract-factory.md:
```diff
@@ -1,67 +1,80 @@
+## Alternatives
 
-## References
+It is also possible to use the [Config Abstract Factory](config-abstract-factory.md), which gives slightly more flexibility in terms of mapping dependencies:
```
```suggestion
It is also possible to use the [Config Abstract Factory](config-abstract-factory.md), which is slightly more flexibility in terms of mapping dependencies:
```

froschdesign committed:
Update documentation for Reflection Abstract Factory

The documentation for Reflection Abstract Factory has been updated to improve the clarity and flow of the information. The changes include rewording for better understanding, reordering of sections for better logical progression, and additional warnings and recommendations concerning performance issues and alternatives.

Signed-off-by: Frank Brückner <dev@froschdesignstudio.de>

settermjd commented on docs/book/v4/reflection-abstract-factory.md:
```diff
@@ -1,67 +1,80 @@
+- Scalar parameters will result in the factory raising an exception, unless a default value is present; if it is, that value will be used.
+- If a service cannot be found for a given typehint, the factory will raise an exception detailing this.
+
+WARNING: `$options` passed to the factory are ignored in all cases, as we cannot make assumptions about which argument(s) they might replace.
```
I don't know if it's applicable, or enabled, but what about [an admonition](https://squidfunk.github.io/mkdocs-material/reference/admonitions/) to draw more attention to this point?

froschdesign replied:
It is an admonition but we do not use Material for MkDocs.
See: https://oprypin.github.io/markdown-callouts/

settermjd commented on docs/book/v4/reflection-abstract-factory.md:
```diff
@@ -1,67 +1,80 @@
-
-You may also use the [Config Abstract Factory](config-abstract-factory.md),
-which gives slightly more flexibility in terms of mapping dependencies:
+TIP:
```
How about formatting this as an admonition as well?

froschdesign replied:
It is.

settermjd reviewed (commented):
Just a bit of feedback for what it's worth.
